File: checksum.cpp
#include <iostream>
#include <fstream>
#include <cstdint>
#include <string>

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <filename>" << std::endl;
        return 1;
    }

    const std::string filename = argv[1];
    const uint32_t RANGE_1_START = 0x0000;
    const uint32_t RANGE_1_END = 0xFEFF;
    const uint32_t RANGE_2_START = 0x10000;
    const uint32_t RANGE_2_END = 0x1FEFF;

    uint32_t checksum1 = 0;
    for (uint32_t i = RANGE_1_START; i <= RANGE_1_END; ++i) {
        std::ifstream file(filename, std::ios::binary);
        file.seekg(i);
        char byte;
        file.get(byte);
        checksum1 += static_cast<uint32_t>(byte);
        file.close();
    }

    uint32_t checksum2 = 0;
    for (uint32_t i = RANGE_2_START; i <= RANGE_2_END; ++i) {
        std::ifstream file(filename, std::ios::binary);
        file.seekg(i);
        char byte;
        file.get(byte);
        checksum2 += static_cast<uint32_t>(byte);
        file.close();
    }

    std::fstream file(filename, std::ios::in | std::ios::out | std::ios::binary);
    if (!file.is_open()) {
        std::cerr << "Failed to open file." << std::endl;
        return 1;
    }

    uint32_t old_checksum1;
    file.seekg(0xFF00);
    file.read(reinterpret_cast<char*>(&old_checksum1), sizeof(old_checksum1));

    uint32_t old_checksum2;
    file.seekg(0x1FF00);
    file.read(reinterpret_cast<char*>(&old_checksum2), sizeof(old_checksum2));

    std::cout << "0x0000-0xFEFF: " << old_checksum1 << std::endl;
    std::cout << "0x10000-0x1FEFF: " << old_checksum2 << std::endl;

    std::string answer;
    std::cout << "Do you want to recalculate the checksum? (y/n): ";
    std::getline(std::cin, answer);

    if (answer == "y") {
        file.seekp(0xFF00);
        file.write(reinterpret_cast<char*>(&checksum1), sizeof(checksum1));

        file.seekp(0x1FF00);
        file.write(reinterpret_cast<char*>(&checksum2), sizeof(checksum2));

        std::cout << "Checksums recalculated." << std::endl;
    } else {
        std::cout << "Checksums not recalculated." << std::endl;
    }

    file.close();

    return 0;
}

File: checksum.py
import struct
import binascii

filename = input("Введите имя файла: ")

ranges = [(0x0000, 0xFEFF), (0x10000, 0x1FEFF)]

for start, end in ranges:
    with open(filename, "rb") as f:
        print(f"Диапазон: 0x{start:04X}-0x{end:04X}:")

        f.seek(start)
        crc = 0
        for byte in bytearray(f.read(end - start + 1)):
            crc += byte
            byte_str = struct.pack('>H', crc & 0xFFFF)

        print(f"  CRC: {binascii.hexlify(byte_str).decode('utf-8')}")

    answer = input("Записать CRC? (y/n): ")
    if answer == "y":
        with open(filename, "r+b") as f:
            f.seek(0xFF00 if start == 0x0000 else 0x1FF00)
            f.write(byte_str)
            print("Успешно")
    else:
        print("Неудачно")
